exercises/variable-length-quantity/.meta/gen.go

```diff
@@ -0,0 +1,146 @@
+// +build ignore
+
+package main
+
+import (
+	"log"
+	"text/template"
+
+	"../../../gen"
+)
+
+func main() {
+	t := template.New("").Funcs(template.FuncMap{
+		"GroupComment": GroupComment,
+		"iByteSlice":   iByteSlice,
+		"byteSlice":    byteSlice,
+		"iUint32Slice": iUint32Slice,
+		"lengthSlice":  lengthSlice,
+	})
+	t, err := t.Parse(tmpl)
+	if err != nil {
+		log.Fatal(err)
+	}
+	var j js
+	if err := gen.Gen("variable-length-quantity", &j, t); err != nil {
+		log.Fatal(err)
+	}
+}
+
+// iByteSlice converts a slice of interface{} values whose
+// underlying members should be float64(JSON decoded integers)
+// to a slice of byte.
+func iByteSlice(ns []interface{}) []byte {
+	b := make([]byte, len(ns))
+	for i, n := range ns {
+		if fn, ok := n.(float64); ok {
+			b[i] = byte(int(fn))
+		}
+	}
+	return b
+}
+
+// byteSlice converts a slice of uint32 to a byte slice.
+func byteSlice(ns []uint32) []byte {
+	b := make([]byte, len(ns))
+	for i, n := range ns {
+		b[i] = byte(n)
+	}
+	return b
+}
+
+// lengthSlice returns the length of given slice.
+func lengthSlice(ns []uint32) int {
+	return len(ns)
+}
+
+// iUint32Slice converts an interface{} to a slice
+// of uint32 values or an empty slice.
+func iUint32Slice(ns interface{}) []uint32 {
+	is, ok := ns.([]interface{})
+	if !ok {
+		return []uint32{}
+	}
+	u := make([]uint32, len(is))
+	for i, n := range is {
+		if un, ok := n.(float64); ok {
+			u[i] = uint32(un)
+		}
+	}
+	return u
+}
+
+// The JSON structure we expect to be able to unmarshal into
+type js struct {
+	Groups []TestGroup `json:"Cases"`
+}
+
+type TestGroup struct {
+	Description string
+	Cases       []OneCase
+}
+
+type OneCase struct {
+	Description string
+	Property    string      // "encode" or "decode"
+	Input       []uint32    // supports both []byte and []uint32 in JSON.
+	Expected    interface{} // supports []byte, []uint32, or null in JSON.
```

Comment:
> Do you know the rational for the new way in the JSON ?

I didn't know off the top of my head, so I went searching. The clearest explanation is probably: https://github.com/exercism/x-common/issues/336#issuecomment-279880708 : 

> In most of the test suites with more than one type of test, the test's type is encoded
in a property-key, in an object describing a test group.
> 
> There are two problem with that approach:
>
> * It mixes two different concepts regarding the tests: grouping and identification
> * It doesn't allow nesting of test groups. That would be nice to have, but is not really needed.
> * It doesn't allow grouping of test of different types, which would be really great.
> 
> Moving the **test type** near the **test data**, we solved all the above problem easily. It is theoretically sound and adds functionality.

I wonder if it would be too much to ask to have the keys again.

I'm wondering what other statically-typed language has generators. There's https://github.com/exercism/xscala/tree/master/testgen/src/main/scala , might need to take a look at what they're doing.